### ISSUE TITLE
New Rule: no-override-native-methods | no-assignment-native-props?

### ISSUE THREAD

**Issue by seahindeniz** [eslint/eslint#15121]
### Rule details

Prevents overriding native props/methods

### Related ECMAScript feature

No

### What type of rule is this?

Warns about a potential problem

### Example code

```js
Number.isNaN = () => {}; // should be prohibited
Object.entries = () => {}; // should be prohibited
console.log = () => {}; // considered as good, it should be prohibited as well


Number.prototype.isNaN = () => {}; // Number prototype is read only, properties should not be added. (no-extend-native)
```

### Why should this rule be in the core instead of a plugin?

Basically, we need to prevent overriding methods or properties of native classes, and I couldn't find a solution that makes it possible. 
Such change might cause pages to work unexpectedly in browsers when there is a native class method get's overridden and used by libraries that are not related to each other. ESLint can alert if that happens

### Participation

- [ ] I am willing to submit a pull request to implement this rule.

### Additional comments

Related discussion https://github.com/eslint/eslint/discussions/15110

**Comment by ljharb** [comment]
https://eslint.org/docs/rules/no-global-assign

**Comment by seahindeniz** [comment]
@ljharb it's not that I think, [here](https://eslint.org/demo#eyJ0ZXh0IjoiTnVtYmVyLmlzTmFOID0gKCkgPT4ge307XG5PYmplY3QuZW50cmllcyA9ICgpID0+IHt9O1xuY29uc29sZS5sb2cgPSAoKSA9PiB7fTsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjEyLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7Im5vLWdsb2JhbC1hc3NpZ24iOjJ9LCJlbnYiOnsiYnJvd3NlciI6dHJ1ZSwibm9kZSI6dHJ1ZX19fQ==)

**Comment by ljharb** [comment]
ah, you're right, https://eslint.org/docs/rules/no-extend-native then?

**Comment by aladdin-add** [comment]
Per our [rules policies](https://eslint.org/blog/2020/05/changes-to-rules-policies), we can only accept new rules if they relate to new ECMAScript features that reached stage 4 in the preceding 12 months, so this proposal doesn't qualify.

I'm in favor of a new option for `no-extend-native`.

**Comment by ljharb** [comment]
I’m surprised no-extend-native doesn’t cover this by default; an option would be great - the airbnb config would absolutely use it.

**Comment by seahindeniz** [comment]
I'm fine with having a new rule or changing an existing rule.

> the third isn't a language builtin though.

I'm not sure how to respond to this properly but from what I can see, console methods are considered as native functions in browser and Node environments. So I think it should be considered in order to maintain integrity and security of the console.X methods.
Can there be some exceptions for the objects that can be evaluated with the "env" property? So that "console" can be ignored in different environments. (just an idea)

**Comment by nzakas** [comment]
no-extend-native is explicitly about adding things to prototypes:

> Disallows directly modifying the prototype of builtin objects.

I don’t think it makes sense to extend to this use case.

I also don’t think this rule is critical to be in the core and would rather see it as a plugin.

**Comment by mdjermanovic** [comment]
I wouldn't mind having this rule in the core. It feels that this rule is missing - we have core rules that disallow replacing `Array` and `Array.prototype.forEach` but we don't have a rule that would disallow replacing `Array.from`.

**Comment by ljharb** [comment]
The name of the rule is "extend native". It's only the description - something that can be changed in a semver-patch - that constrains it to prototypes.

It's very very critical to do whatever is possible to prevent developers from mutating any builtins, as this brilliant author makes clear http://www.nczonline.net/blog/2010/03/02/maintainable-javascript-dont-modify-objects-you-down-own/ ;-)

**Comment by nzakas** [comment]
@ljharb “extend” means adding to, not overwriting. These are two different cases and “extending” this rule for a different case feels like trying to find a loophole around our new rule policy. 

Aside from not fitting our criteria for new rules, this rule flags an obvious intentional bad practice - the people doing it know what they are doing; if we were to make an exception (which I would prefer to never do), it should be for an unintentional bad practice. I’d be more convinced if we were protecting someone from something they didn’t know they were doing.

**Comment by ljharb** [comment]
So if i do `Number.foo = 3` it will warn?

**Comment by seahindeniz** [comment]
_Excuse me if my words make you guys feel uncomfortable or offensive in any way. I apologize in advance, and no offense as well._

I have no deep knowledge about the current policies of ESLint. By simply looking at this case, the criteria for adding new core rules need to be re-evaluated to have new aspects for the forgotten things. I'm saying this because, in my opinion, ESLint hasn't finalized its core rules yet- especially for the one that is about accepting only new ES features and ignore other suggestions.

Also, I believe, it is out of concern that ESLint shouldn't care about people's intentions. In fact, it should care about projects and the health of the environment of the project that runs in. Since ESLint has rules about preventing devs from extending or overriding the Native objects, the suggestion is as important as overriding the native objects, and it should care about the expected outcomes and concerns and warn accordingly.

In small projects or projects that don't have CI flows for ESLint, overriding `Number.isNaN` can happen intentionally, or some other situations may occur. But on the other hand, override should be checked always to make sure that the project is secure for its environment and no pollution will occur in the future

**Comment by nzakas** [comment]
@seahindeniz we realize that our policy isn’t what everyone would like, but that’s why ESLint supports plugins. We never want to stand in the way of people having the rules that they want, but we just can’t keep adding new rules to the core forever. With about 300 rules already, it’s a maintenance nightmare. 

For more information, please see the blog post where we announced this change:
https://eslint.org/blog/2020/05/changes-to-rules-policies

> In small projects or projects that don't have CI flows for ESLint, overriding Number.isNaN can happen intentionally, or some other situations may occur.

It’s clear that you feel strongly about this, and that’s why we have custom rules. You are still free to create a rule that does this, and to use it and even publicize why it’s important to use. This is how ESLint was intended to work: you can configure and extend it to do exactly what you want.

**Comment by nzakas** [comment]
We aren’t moving forward here, so closing.

**Comment by ljharb** [comment]
@nzakas to be clear, extend means both adding to and overwriting (in class inheritance terminology); i do think this falls under the scope of what the rule claims to cover.
